docs/runtime/runtime-metadata.md

```diff
@@ -0,0 +1,196 @@
+
+```rust
+pub struct RuntimeMetadataV4 {
+    pub modules: DecodeDifferentArray<ModuleMetadata>,
+}
+```
+
+As you can see, the overall runtime's metadata is an array of module's metadata. These modules includes predefined SRML and your custom runtime modules. Each runtime module is mainly composed by storage entries, dispatchable functions and events.
+
+```rust
+pub struct ModuleMetadata {
+    pub name: DecodeDifferentStr,
+    pub prefix: DecodeDifferent<FnEncode<&'static str>, StringBuf>,
+    pub storage: ODFnA<StorageFunctionMetadata>,
+    pub calls: ODFnA<FunctionMetadata>,
+    pub event: ODFnA<EventMetadata>,
+}
+```
+
+Let's dig into the detail:
+* `name`: a string shows module's name.
+* `prefix`: an encodable function that produces store's metadata name.
+* `storage`: an array of storage entry's metadata which described in `StorageFunctionMetadata`. 
+* `calls`: an array of dispatchable function's metadata which described in `FunctionMetadata`.
+* `event`: an array of event's metadata which described in `EventMetadata`.
+
+```rust
+// The metadata about one storage entry
+pub struct StorageFunctionMetadata {
+    // The name of the storage entry
+    pub name: DecodeDifferentStr,
+    // The storage entry modifier: Optional or Default
+    pub modifier: StorageFunctionModifier,
+    // The storage entry type: Plain, Map or DoubleMap
+    pub ty: StorageFunctionType,
+    // The initiated byte value for this entry
+    pub default: ByteGetter,
+    // The documents of the storage entry
+    pub documentation: DecodeDifferentArray<&'static str, StringBuf>,
+}
+
+// The metadata about a function
+pub struct FunctionMetadata {
+    // The name of the function
+    pub name: DecodeDifferentStr,
+    // The arguments in function's definition, each includes parameter name and type.
+    pub arguments: DecodeDifferentArray<FunctionArgumentMetadata>,
+    // The documents of the function
+    pub documentation: DecodeDifferentArray<&'static str, StringBuf>,
+}
+
+// The metadata about an event
+pub struct EventMetadata {
+    // The name of the event
+    pub name: DecodeDifferentStr,
+    // The arguments of the event
+    pub arguments: DecodeDifferentArray<&'static str, StringBuf>,
+    // The documents of the event
+    pub documentation: DecodeDifferentArray<&'static str, StringBuf>,
+}
+```
+
+Substrate framework then prefixes the runtime metadata with a hard coded `u32` for reserved usage, and encodes to `Vec<u8>` blob for transferring.
+
+## How to Use The Metadata
+
+The metadata of runtime is consolidated and encoded automatically when developing runtime modules. There are a few scenarios that the metadata can really help you out:
+
+* You want to interact with a Substrate blockchain, but don't have much idea about what features the blockchain provides.
```

Comment:
This should be the beginning of the article. Note that it's not _How to Use the Metadata_, but rather how the metadata is relevant.

Reply:
In the beginning, we already have this statement:
```
Substrate module-based metadata provides an overview of blockchain runtime. It helps external code introspect modules, storage items, dispatchable functions and events.
```
This scenario-based introduction can be meaningful when the user does want to use the metadata.

> Note that it's not How to Use the Metadata, but rather how the metadata is relevant.

`How the metadata is relevant` is so generic that makes user confused about what section really about. My point here is after user know what metadata is, I want to teach them how to use it.

Reply:
> How the metadata is relevant is so generic that makes user confused about what section really about.

Explaining why something is relevant belongs at the beginning of the article in the introduction. I wasn't suggesting a section heading.

> My point here is after user know what metadata is, I want to teach them how to use it.

But these statements don't teach them how to use it.

You should eliminate the part about explaining what a runtime is and what metadata does not do and then add these scenarios where you would want to interact with metadata.